### PR TITLE
fix: add minimum requirements for gcp-releasetool

### PR DESCRIPTION
**Pull request by parthea**
BEGIN_COMMIT_OVERRIDE
fix!: Require cryptography 42.x.x
feat: Add support for python 3.12
fix: Require python 3.8+
fix: Drop usage of `pkg_resources`
fix: Drop usage of `pkgutil`
fix: Update dependencies in requirements.txt and requirements-dev.txt
fix: Add minimum requirements for gcp-releasetool
ci: Updates the `test` nox session to test the minimum version of dependencies via `testing/constraints-3.7.txt`
ci: Error on deprecation warnings
END_COMMIT_OVERRIDE

Fixes https://github.com/googleapis/releasetool/issues/533

### CODE REVIEW

File: releasetool/github.py [hunk 1]
```diff
@@ -22,7 +22,15 @@
...
-from cryptography.hazmat.backends import default_backend
+
+import cryptography
+
+CRYPTOGRAPHY_VERSION = cryptography.__version__
+
+if int(CRYPTOGRAPHY_VERSION[0:2]) < 42:
+    from cryptography.hazmat.backends import default_backend
+else:
+    from cryptography.hazmat.primitives import serialization
```
Comment (chingor13) [inline releasetool/github.py:33]: Do we really want to try to maintain support for multiple versions of cryptography?

File: releasetool/update_check.py [hunk 2]
```diff
@@ -14,11 +14,16 @@
...
+import sys
 
-import pkg_resources
 import packaging.version
 import requests
 
+if sys.version_info < (3, 8):
+    import importlib_metadata as metadata
+else:
+    import importlib.metadata as metadata
```
Comment (chingor13) [inline releasetool/update_check.py:25]: Consider dropping python 3.7 which is EoL.

**Review by chingor13 — CHANGES_REQUESTED** [review]
This is an internal tool, so we can make breaking changes as necessary assuming we still version correctly and our usage is locking versions.

If necessary, we can backport fixes to our previous semver major.

**Comment by chingor13** [comment]
Note that the tests are breaking probably because of our global pip install in https://github.com/googleapis/releasetool/blob/master/.kokoro/build.sh

nox is trying to maintain sessions, but I think a version of something is leaking through.

File: releasetool/update_check.py (hunk 2 above)
Comment (parthea) [reply releasetool/update_check.py]: Fixed in 5ea803c2de145a06d828fbc6c93827caba801b2a

File: releasetool/github.py (hunk 1 above)
Comment (parthea) [reply releasetool/github.py]: Fixed in 91c2e4055e5a2b1746d8bd18f7f1a792a20724f9 and 69ddaef56519dd461e5ee4ecde5c7ac9c99ad5e4

File: releasetool/github.py [hunk 3]
```diff
@@ -22,8 +22,11 @@
 
 import jwt
 import requests
-from cryptography.hazmat.backends import default_backend
 
+import cryptography
+from cryptography.hazmat.primitives import serialization
+
+CRYPTOGRAPHY_VERSION = cryptography.__version__
```
Comment (chingor13) [inline releasetool/github.py:29]: This looks unused - can remove this and its import.

**Review by chingor13 — CHANGES_REQUESTED** [review]

File: releasetool/github.py (hunk 3 above)
Comment (parthea) [reply releasetool/github.py]: Fixed in https://github.com/googleapis/releasetool/pull/532/commits/e14375e752a4caa5141f302c4baf21a1c39b6081

File: setup.py [hunk 4]
```diff
@@ -19,18 +19,18 @@
...
-    "protobuf>=3.19.4, <4.0.0",
-    "pyjwt",
-    "pyperclip",
-    "python-dateutil",
+    "cryptography>=42",
+    "google-auth>=1.10.1",
+    "jinja2>=3.1.3",
+    "keyring>=21.8.0",
+    "packaging>=20.0",
+    "protobuf>=4.21.6",
```
Comment (chingor13) [inline setup.py:30]: Can we separate out the protobuf major version bump into a separate PR? This looks like a dangerous change (for potentially breaking the triggering of our builds) and we don't have great tests around it.

If you are confident that the wire format is unchanged, then perhaps we can merge together.

**Review by chingor13 — CHANGES_REQUESTED** [review]
Test is currently failing

File: setup.py (hunk 4 above)
Comment (parthea) [reply setup.py:30]: > If you are confident that the wire format is unchanged, then perhaps we can merge together.

Perhaps we should publish an rc version with these changes and test it out. WDYT?

**Review by chingor13 — APPROVED** [review]

**Pull request merged**